src/libsyntax/ast.rs

```diff
@@ -630,6 +631,9 @@ pub enum PatKind {
     /// `0 <= position <= subpats.len()`.
     TupleStruct(Path, Vec<P<Pat>>, Option<usize>),
 
+    /// An or-pattern `A | B | C`.
+    Or(Vec<P<Pat>>),
```

Comment:
So I would expect us to have:
```rust
pub struct Arm {
    pub hir_id: HirId,
    pub span: Span,
    pub attrs: HirVec<Attribute>,
    pub pat: P<Pat>,  // <-- It is `pub pats: HirVec<P<Pat>>,` right now.
    pub guard: Option<Guard>,
    pub body: P<Expr>,
}
```
after this PR. To do this, [`fn parse_pats`](https://doc.rust-lang.org/nightly/nightly-rustc/src/syntax/parse/parser.rs.html#3502-3528), preferably renamed to `fn parse_top_pat`, should return a `P<Pat>` but allow a leading `|` (first line: `self.eat(&token::BinOp(token::Or));`) and then immediately use a `fn parse_pat_allow_top_alt`.

I think you can follow the grammar rules in the RFC quite literally.